src/components/item/sharing/CategorySelection.js

```diff
@@ -0,0 +1,211 @@
+  );
+
+  const mapItemCategory = new Map(
+    itemCategories?.map((obj) => [obj.category_id, obj.id]),
+  );
+  const ageCategory = ageList?.map((entry) => entry.id);
+  const disciplineCategory = disciplineList?.map((entry) => entry.id);
+
+  // seperate age and discipline category
+  const initialAgeValue = itemCategories?.filter((itemCategory) =>
+    ageCategory?.includes(itemCategory.category_id),
+  );
+  const initialDisciplineValue = itemCategories?.filter((itemCategory) =>
+    disciplineCategory?.includes(itemCategory.category_id),
+  );
+  const selectedDisciplineIds = initialDisciplineValue
+    ?.map((entry) => entry.category_id)
+    .toArray();
+  const selectedAgeIds = initialAgeValue
+    ?.map((entry) => entry.category_id)
+    .toArray();
+  // set initial value
+  const [ageOptionsValue, setAgeOptionsValue] = useState([]);
+  const [disciplineOptionsValue, setDisciplineOptionsValue] = useState([]);
+
+  // update state variables depending on fetch values
+  useEffect(() => {
+    if (selectedAgeIds && ageList) {
+      setAgeOptionsValue(
+        ageList.filter((entry) => selectedAgeIds.includes(entry.id)).toArray(),
+      );
+    }
+  }, [item, itemCategories, ageList, selectedAgeIds]);
+
+  useEffect(() => {
+    if (selectedDisciplineIds && disciplineList) {
+      setDisciplineOptionsValue(
+        disciplineList
+          .filter((entry) => selectedDisciplineIds.includes(entry.id))
+          .toArray(),
+      );
+    }
+  }, [item, itemCategories, disciplineList, selectedDisciplineIds]);
+
+  if (
+    isMemberLoading ||
+    isItemCategoriesLoading ||
+    isCategoryTypesLoading ||
+    isCategoriesLoading
+  ) {
+    return <Loader />;
+  }
+
+  const handleAgeChange = (event, value, reason) => {
+    if (reason === 'select-option') {
+      // post new category (discipline)
+      const newCategoryId = value.at(-1).id;
+      createItemCategory({
+        itemId,
+        categoryId: newCategoryId,
+      });
+    } else if (reason === 'remove-option') {
+      // remove an option
+      const result = ageOptionsValue.filter(
+        ({ id: id1 }) => !value.some(({ id: id2 }) => id2 === id1),
+      );
+      const entryId = mapItemCategory.get(result[0].id);
+      deleteItemCategory({
+        itemId,
+        entryId,
+      });
+    }
+  };
+
+  const handleDisciplineChange = (event, value, reason) => {
+    if (reason === 'select-option') {
+      // post new category (discipline)
+      const newCategoryId = value.at(-1).id;
+      createItemCategory({
+        itemId,
+        categoryId: newCategoryId,
+      });
+    } else if (reason === 'remove-option') {
+      // remove an option
+      const result = disciplineOptionsValue.filter(
+        ({ id: id1 }) => !value.some(({ id: id2 }) => id2 === id1),
+      );
+      const entryId = mapItemCategory.get(result[0].id);
+      deleteItemCategory({
+        itemId,
+        entryId,
+      });
+    }
+  };
```

Comment:
What about a handleChangeBuilder ?

`handleChangeBuilder(mutate, allValues) => (event, value, reason) => {....}`

Reply:
I merged the two handlChange into one, but it's slightly different than what you suggested, because I'm not very sure what you mean by (mutate and allValues)